tests/modules/encoder_base_test.py

```diff
@@ -0,0 +1,161 @@
+# pylint: disable=no-self-use,protected-access
+import numpy
+import torch
+from torch.autograd import Variable
+from torch.nn import LSTM
+
+from allennlp.modules.encoder_base import _EncoderBase
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.nn.util import sort_batch_by_length, get_lengths_from_binary_sequence_mask
+
+
+class TestEncoderBase(AllenNlpTestCase):
+
+    def setUp(self):
+        super(TestEncoderBase, self).setUp()
+        self.lstm = LSTM(bidirectional=True, num_layers=3, input_size=3, hidden_size=7, batch_first=True)
+        self.encoder_base = _EncoderBase(stateful=True)
+
+        tensor = Variable(torch.rand([5, 7, 3]))
+        tensor[1, 6:, :] = 0
+        tensor[2, :, :] = 0  # <= completely masked
```

Comment:
Related to some of my comments above - if you were guaranteed that all of these would be at the end, would it make it possible to simplify the handling of the `num_valid` stuff?

Reply:
Do you mean for the test, or for the module? We should have this not at the end in the test to check that the sorting is happening correctly. We can't take only sorted states in the module, because if we did that it would be very difficult to keep track of the state, as it would have to be stored in a sorted format, which seems unpleasant and very error prone.

Reply:
I meant the padding, not the sorting.  If you're guaranteed that all invalid instances are at the end, can you simplify the logic so that `sort_batch_by_length` can return something of size `(num_valid, ...)` instead of `(batch_size, ...)`?  Not a super important question.

Reply:
I removed it, the test still passes 👍